MapboxNavigation/InstructionsCardCollection.swift

```diff
@@ -0,0 +1,332 @@
+import MapboxDirections
+import MapboxCoreNavigation
+
+@objc public protocol InstructionsCardCollectionDelegate: InstructionsCardContainerViewDelegate {
+    /**
+     Called when previewing the steps on the current route.
+     
+     Implementing this method will allow developers to move focus to the maneuver that corresponds to the step currently previewed.
+     - parameter instructionsCardCollection: The instructions card collection instance.
+     - parameter step: The step for the maneuver instruction in preview.
+     */
+    @objc(instructionsCardCollection:previewForStep:)
+    func instructionsCardCollection(_ instructionsCardCollection: InstructionsCardCollection, previewFor step: RouteStep)
+    
+    /**
+     Offers the delegate the opportunity to customize the size of a prototype collection view cell per the associated trait collection.
+     
+     - parameter instructionsCardCollection: The instructions card collection instance.
+     - parameter traitCollection: The traitCollection associated to the current container view controller.
+     - returns: The preferred size of the cards for each cell in the instructions card collection.
+     */
+    @objc(instructionsCardCollection:cardSizeForTraitCollection:)
+    optional func instructionsCardCollection(_ instructionsCardCollection: InstructionsCardCollection, cardSizeFor traitCollection: UITraitCollection) -> CGSize
+}
+
+open class InstructionsCardCollection: UIViewController {
+    typealias InstructionsCardCollectionLayout = UICollectionViewFlowLayout
+    
+    var routeProgress: RouteProgress?
+    var cardSize: CGSize = .zero
+    var cardStyle: DayInstructionsCardStyle = DayInstructionsCardStyle()
+    
+    var instructionCollectionView: UICollectionView!
+    var instructionsCardLayout: InstructionsCardCollectionLayout!
+    
+    public private(set) var isInPreview = false
+    private var currentStepIndex: Int?
+    
+    var steps: [RouteStep]? {
+        guard let stepIndex = routeProgress?.currentLegProgress.stepIndex, let steps = routeProgress?.currentLeg.steps else { return nil }
+        var mutatedSteps = steps
+        if mutatedSteps.count > 1 {
+            mutatedSteps = Array(mutatedSteps.suffix(from: stepIndex))
+            mutatedSteps.removeLast()
+        }
+        return mutatedSteps
+    }
+    
+    var distancesFromCurrentLocationToManeuver: [CLLocationDistance]? {
```

Comment:
This property as well as `step` appear to be strictly related to the object model – they don’t use anything else in self. Can we give them more general-purpose names and move them to RouteProgress or an extension of that class? That’ll make it much easier to unit-test this functionality.

Reply:
Eventually, the goal is to make this a data source provided to the instructions card. This `InstructionsCardCollection` class would not contain any information related to how the distances displayed on each card is derived.